File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/MatchResult.java
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public record MatchResult(
        @NotNull String value,
        int index,
        @Nullable String name
) {
    public MatchResult(@NotNull String value, int index) {
        this(value, index, null);
    }
    @Contract("_ -> new")
    public @NotNull MatchResult withOffset(int offset) {
        return new MatchResult(value, index + offset, name);
    }

    @Contract("_ -> new")
    public @NotNull MatchResult withIndex(int index) {
        return new MatchResult(value, index, name);
    }
}

File: src/main/kotlin/io/github/colintimbarndt/chat_emotes/data/unicode/pattern/MatchResult.kt
package io.github.colintimbarndt.chat_emotes.data.unicode.pattern

/**
 * Represents the result of a [Matcher]
 */
data class MatchResult(
    /**
     * The sequence of characters that was matched
     */
    val value: String,
    /**
     * The unique index of the matched sequence
     * @see Matcher
     */
    val index: Int,
    /**
     * A name that was assigned to the matched sequence by the [Matcher], which is optional
     */
    val name: String? = null
) {
    /**
     * @return New [MatchResult] with [offset] added to the [index]
     */
    infix fun offsetBy(offset: Int): MatchResult {
        return MatchResult(value, index + offset, name)
    }

    /**
     * @return New [MatchResult] with the given index
     */
    infix fun withIndex(index: Int): MatchResult {
        return MatchResult(value, index, name)
    }
}